core/templates/pages/exploration-player-page/services/exploration-engine.service.spec.ts

```diff
@@ -0,0 +1,910 @@
+    }).toThrowError(
+      'Cannot read property \'getAuthorRecommendedExpIds\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    explorationEngineService.currentStateName = 'End';
+
+    const recommendedId = explorationEngineService.getAuthorRecommendedExpIds();
+    expect(recommendedId).toContain('recommnendedExplorationId');
+  });
+
+  it('should update current state when an answer is submitted ' +
+    'and a new card is recorded', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+    let submitAnswerSuccessCb = jasmine.createSpy('success');
+    let answer = 'answer';
+    let answerClassificationResult = new AnswerClassificationResult(
+      outcomeObjectFactory.createFromBackendDict({
+        dest: 'Mid',
+        feedback: {
+          content_id: 'feedback_1',
+          html: 'Answer is correct!'
+        },
+        labelled_as_correct: true,
+        param_changes: [],
+        refresher_exploration_id: null,
+        missing_prerequisite_skill_id: null
+      }), 1, 0, 'default_outcome');
+
+    let lastCard = stateCardObjectFactory.createNewCard(
+      'Card 1', 'Content html', 'Interaction text', null,
+      null, null, 'content_id');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+    spyOn(playerTranscriptService, 'getLastStateName')
+      .and.returnValue('Start');
+    spyOn(playerTranscriptService, 'getLastCard').and.returnValue(lastCard);
+    spyOn(answerClassificationService, 'getMatchingClassificationResult')
+      .and.returnValue(answerClassificationResult);
+
+    expect(explorationEngineService.currentStateName).toBe(undefined);
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    explorationEngineService.submitAnswer(
+      answer, textInputService, submitAnswerSuccessCb);
+    expect(explorationEngineService.currentStateName).toBe('Start');
+    explorationEngineService.recordNewCardAdded();
+    expect(explorationEngineService.currentStateName).toBe('Mid');
+  });
+
+  it('should load initial state when moved to new exploration', () => {
+    let moveToExplorationCb = jasmine.createSpy('success');
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    explorationEngineService.exploration = explorationObjectFactory
+      .createFromBackendDict(explorationDict);
+
+    let currentStateName = explorationEngineService.currentStateName;
+    expect(currentStateName).toBe(undefined);
+
+    // Please note that we are not calling init funtion here.
+    explorationEngineService.moveToExploration(moveToExplorationCb);
+
+    currentStateName = explorationEngineService.currentStateName;
+    let initalState = explorationEngineService.exploration.initStateName;
+    expect(currentStateName).toBe(initalState);
+  });
+
+  it('should return true if current state is initial state', () => {
```

Comment:
Ok, the final error log would read "Exploration Engine Service should return true if current state is initial state" which is not right. (Similar to above test names). Can we make the naming slightly better?

Reply:
changed here and elsewhere.